File: day12_2/src/main.rs
use std::collections::{BTreeMap, HashSet};

use aoc::aoc;
use itertools::Itertools;

#[aoc(2024, 12, 2)]
fn main(input: &str) -> usize {
    let garden = input
        .lines()
        .enumerate()
        .flat_map(|(y, line)| {
            line.trim()
                .chars()
                .enumerate()
                .map(move |(x, label)| ((x as isize, y as isize), label))
        })
        .collect::<Garden>();

    let mut remaining_plots = garden.clone();
    let mut result = 0;

    while !remaining_plots.is_empty() {
        let (&plot, &label) = remaining_plots.iter().next().unwrap();
        let region = find_region(plot, label, &garden);
        let area = region.len();
        let perimeter = perimeter(&region, &garden);
        let sides = sides(&perimeter, &garden);
        let price = area * sides;

        result += price;

        for plot in region {
            remaining_plots.remove(&plot);
        }
    }

    result
}

type Plot = (isize, isize);
type Garden = BTreeMap<Plot, char>;
type Region = HashSet<Plot>;
type Label = char;

fn find_region(plot: Plot, label: Label, garden: &Garden) -> Region {
    let mut region = HashSet::new();

    find_region_rec(plot, &mut region, label, garden);

    region
}

fn find_region_rec(plot: Plot, region: &mut Region, label: Label, garden: &Garden) {
    if region.contains(&plot) {
        return;
    }

    if garden.get(&plot) != Some(&label) {
        return;
    }

    region.insert(plot);

    for neighbour in neighbours(plot) {
        find_region_rec(neighbour, region, label, garden);
    }
}

fn perimeter(region: &Region, garden: &Garden) -> HashSet<Plot> {
    let plot = region.iter().next().unwrap();
    let label = garden.get(plot).unwrap();
    let mut perimeter = HashSet::new();

    for plot in region {
        for neighbour in neighbours(*plot) {
            if garden.get(&neighbour) == Some(label) {
                continue;
            }

            perimeter.insert(*plot);
        }
    }

    perimeter
}

fn sides(perimeter: &Region, garden: &Garden) -> usize {
    let plot = perimeter.iter().next().unwrap();
    let label = *garden.get(plot).unwrap();

    perimeter
        .iter()
        .flat_map(|&plot| {
            neighbours_with_directions(plot)
                .filter(|(neighbour, _)| {
                    garden
                        .get(neighbour)
                        .is_none_or(|neighbour_label| label != *neighbour_label)
                })
                .map(move |(_, dir)| Side { plot, dir })
        })
        .sorted_by(|a, b| {
            a.dir
                .cmp(&b.dir)
                .then(a.cross_axis_coord().cmp(&b.cross_axis_coord()))
                .then(a.axis_coord().cmp(&b.axis_coord()))
        })
        .coalesce(|a, b| {
            if a.dir != b.dir {
                return Err((a, b));
            }

            if a.cross_axis_coord() != b.cross_axis_coord() {
                return Err((a, b));
            }

            if a.axis_coord() + 1 != b.axis_coord() {
                return Err((a, b));
            }

            Ok(b)
        })
        .count()
}

fn neighbours((x, y): Plot) -> impl Iterator<Item = Plot> {
    [
        (x, y - 1), // up
        (x, y + 1), // down
        (x - 1, y), // left
        (x + 1, y), // right
    ]
    .into_iter()
}

fn neighbours_with_directions((x, y): Plot) -> impl Iterator<Item = (Plot, Dir)> {
    [
        ((x, y - 1), Dir::Up),
        ((x, y + 1), Dir::Down),
        ((x - 1, y), Dir::Left),
        ((x + 1, y), Dir::Right),
    ]
    .into_iter()
}

#[derive(Debug)]
struct Side {
    plot: Plot,
    dir: Dir,
}

impl Side {
    fn axis_coord(&self) -> isize {
        self.dir.axis().get_coord(self.plot)
    }

    fn cross_axis_coord(&self) -> isize {
        self.dir.cross_axis().get_coord(self.plot)
    }
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
enum Dir {
    Up,
    Down,
    Left,
    Right,
}

impl Dir {
    fn axis(&self) -> Axis {
        match self {
            Dir::Up | Dir::Down => Axis::Vertical,
            Dir::Left | Dir::Right => Axis::Horizontal,
        }
    }

    fn cross_axis(&self) -> Axis {
        match self {
            Dir::Up | Dir::Down => Axis::Horizontal,
            Dir::Left | Dir::Right => Axis::Vertical,
        }
    }
}

enum Axis {
    Horizontal,
    Vertical,
}

impl Axis {
    fn get_coord(&self, (x, y): Plot) -> isize {
        match self {
            Axis::Horizontal => y,
            Axis::Vertical => x,
        }
    }
}

File: day12_2/py/main.py
from dataclasses import dataclass
from enum import Enum, IntEnum, auto
from operator import attrgetter, methodcaller
from typing import Dict, Generator, List, Set, Tuple

input = open("input.txt").read()

type Plot = Tuple[int, int]
type Label = str

def main() -> None:
    garden = Garden.parse(input)
    regions = garden.regions()

    print(f"part 1: {part1(regions)}")
    print(f"part 2: {part2(regions)}")

def part1(regions: List["Region"]) -> int:
    total = 0

    for region in regions:
        area = len(region.plots)
        perimeter = len(region.perimeter())
        price = area * perimeter
        total += price

    return total

def part2(regions: List["Region"]) -> int:
    total = 0

    for region in regions:
        area = len(region.plots)
        sides = len(region.edges())
        price = area * sides
        total += price

    return total

@dataclass
class Garden:
    plot_labels: Dict[Plot, Label]
    
    @staticmethod
    def parse(garden: str) -> "Garden":
        plot_labels = {}

        for y, line in enumerate(garden.splitlines()):
            for x, label in enumerate(line):
                plot_labels[(x, y)] = label

        return Garden(plot_labels)
    
    def regions(self) -> List["Region"]:
        remaining = dict(self.plot_labels)
        regions = []

        while remaining:
            plot, label = next(iter(remaining.items()))
            region = self.region(plot, label)

            regions.append(region)

            for plot in region.plots:
                remaining.pop(plot)
            
        
        return regions
    
    def region(self, plot: Plot, label: Label) -> "Region":
        plots: Set[Plot] = set()
        to_check: Set[Plot] = set([plot])

        while to_check:
            plot = to_check.pop()

            if plot in plots:
                continue

            if self.plot_labels.get(plot) != label:
                continue

            plots.add(plot)

            for neighbour in neighbours(plot): 
                to_check.add(neighbour)

        return Region(label, plots)

@dataclass
class Region:
    label: str
    plots: Set[Plot]

    def perimeter(self) -> List["Edge"]:
        edges = []

        for plot in self.plots:
            for neighbour, dir in neighbours_with_dirs(plot):
                if neighbour not in self.plots:
                    edges.append(Edge(dir, plot))

        return edges
    
    def edges(self) -> List[List["Edge"]]:
        def inner() -> Generator[List[Edge]]:
            edges = self.perimeter()
            edges = sorted(edges, key = methodcaller('axis_coord'))
            edges = sorted(edges, key = methodcaller('cross_axis_coord'))
            edges = sorted(edges, key = attrgetter('dir'))

            side: List[Edge] = []

            for edge in edges:
                if not side:
                    side.append(edge)
                    continue

                last = side[-1];
                dirs_match = last.dir == edge.dir
                cross_axes_match = last.cross_axis_coord() == edge.cross_axis_coord()
                axes_consecutive = last.axis_coord() + 1 == edge.axis_coord()

                if all([dirs_match, cross_axes_match, axes_consecutive]):
                    side.append(edge)
                else:
                    yield side
                    side = [edge]

            yield side
        return list(inner())

@dataclass
class Edge:
    dir: "Dir"
    plot: Plot

    def axis_coord(self) -> int:
        return self.dir.axis().get_coord(self.plot)

    def cross_axis_coord(self) -> int:
        return self.dir.cross_axis().get_coord(self.plot)

class Dir(IntEnum):
    Up = auto()
    Down = auto()
    Left = auto()
    Right = auto()

    def axis(self) -> "Axis":
        match self:
            case Dir.Up | Dir.Down: return Axis.Vertical
            case Dir.Left | Dir.Right: return Axis.Horizontal
    
    def cross_axis(self) -> "Axis":
        match self:
            case Dir.Up | Dir.Down: return Axis.Horizontal
            case Dir.Left | Dir.Right: return Axis.Vertical

class Axis(Enum):
    Horizontal = auto()
    Vertical = auto()

    def get_coord(self, plot: Plot) -> int:
        x, y = plot

        match self:
            case Axis.Horizontal: return y
            case Axis.Vertical: return x

def neighbours(plot: Plot) -> List[Plot]:
    return list(neighbour for neighbour, _ in neighbours_with_dirs(plot))

def neighbours_with_dirs(plot: Plot) -> List[Tuple[Plot, Dir]]:
    x, y = plot

    return [
        ((x, y -1), Dir.Up),
        ((x, y + 1), Dir.Down),
        ((x - 1, y), Dir.Left),
        ((x + 1, y), Dir.Right),
    ]

if __name__ == "__main__":
    main()
